### PR TITLE
Pop 21.10: Major R2 Auto Login Bug

### PR DESCRIPTION
**Pull request by 13r0ck**
In a rare "race-like" condition that only occurs on some hardware, cosmic-dock can be loaded before the monitor is ready to display, thus cosmic-dock runs in a "headless" mode, and `mainDock` returns null. Because cosmic-workspaces is loaded after cosmic-dock, this causes the dock to be functional, but the rest of the UI to not be. 

This fixes assembly's `Pop 21.10: Major R2 Auto Login Bug with AMD GPUs, especially 6600xt` issue.

### CODE REVIEW

File: workspace.js
Comment (jackpot51) [inline workspace.js]: Will this cause the fix from https://github.com/pop-os/cosmic-workspaces/pull/36 to not work in this case?

Comment (13r0ck) [reply workspace.js:107]: Aw man. Good catch, I haven't tested it yet, but you are almost certainly correct.

Do you think this translation logic should get moved out of the init and into the render loop? Maybe [here](https://github.com/pop-os/cosmic-workspaces/blob/master_impish/workspacesView.js#L104)?

Comment (13r0ck) [reply workspace.js]: Aw man. Yep you are correct. I fixed the commit to make the translation in the render loop rather than the init

**Comment by 13r0ck** [comment]
This now also fixes the [nitpick](https://github.com/pop-os/cosmic-workspaces/issues/4#issuecomment-964545373) from #36.

**Review by jackpot51 — DISMISSED** [review]

**Comment by jacobgkau** [comment]
Before, when the dock was placed on the left side of the screen, windows popped to the left at the end of the workspace selection animation:

https://user-images.githubusercontent.com/7199422/151250678-749c3cba-fef8-440a-8b3a-18bb84a65db1.mp4


Now, they pop to the right instead:

https://user-images.githubusercontent.com/7199422/151250701-7f2d902c-7bed-4e67-9940-34a356c5149c.mp4


Not exactly a regression, but I thought I'd mention it.

**Comment by 13r0ck** [comment]
Good point! @jacobgkau , I don't like that they pop at all I will look into fixing that in another PR. Thanks!

**Comment by 13r0ck** [comment]
# Please don't merge just yet. I still have one more test on the Major.

**Comment by jacobgkau** [comment]
I'm also not seeing this fix the [nitpick](https://github.com/pop-os/cosmic-workspaces/issues/4#issuecomment-964545373) that you mentioned; the close button still overlaps the workspace picker in that configuration.

Screenshot from 2022-01-26 14-44-55

**Comment by 13r0ck** [comment]
Thank you for the catch again. @jacobgkau  how about now?

**Comment by jacobgkau** [comment]
> Thank you for the catch again. @jacobgkau how about now?

Yes! Close button overlap is fixed on b02440b.

**Review by jackpot51 — DISMISSED** [review]

**Comment by jacobgkau** [comment]
@13r0ck reported Major R2 auto-login is still broken with b02440b.

I'd be inclined to still merge it for the close button fix, @13r0ck let me know if there's anything else you want to try for auto-login on this branch before we move forward with it.

**Comment by 13r0ck** [comment]
With some minor improvements this might fix #35 too. There are other places in cosmic workspaces where the dock is called blindly. 

As for the auto login bug. I'm 80% sure it's upstream gnome-shell, and/or an issue in our extension causing issues in gnome shell. I need to gauge how long that will take to fix, if too long then we can merge this, but i would prefer to fix both at the same time.

**Comment by 13r0ck** [comment]
The issue could be because there is a delay in when the monitor is detected, which then runs gnome-shell in headless mode for a brief moment, but mutter doesn't support headless rendering yet.  Just a theory at the moment

**Comment by 13r0ck** [comment]
@jacobgkau Yep it was a gnome-shell bug. Fixed in pop-os/gnome-shell#82. These two should be tested and merged together.

**Review by jackpot51 — DISMISSED** [review]

**Comment by 13r0ck** [comment]
I forgot to test disableing/uninstalling cosmic-dock. I'm stills comfortable getting this merged, but it should probably be checked first

**Review by jacobgkau — DISMISSED** [review]
https://github.com/pop-os/cosmic-workspaces/issues/4 still isn't occurring. Close button overlap is still fixed with picker & dock on the right side, although it's still occurring with workspace picker on the right and dock at the bottom (this is not a regression):

Screenshot from 2022-01-28 12-41-09

This also fixed an issue with the windows overlapping the workspace picker a bit more when picker + dock are right-aligned and the dock is set to auto-hide. When Workspaces is opened and the dock shows, the windows now dodge the dock along with the workspace picker until the dock hides again.

Tested with https://github.com/pop-os/gnome-shell/pull/82. I did also try disabling cosmic-dock, the behavior is the same as before (Workspaces menu breaks until the shell is restarted, but after that, it works fine.)

**Comment by 13r0ck** [comment]
> Screenshot from 2022-01-28 12-41-09

Oh that's an easy fix, I didn't realize that was also broken, @jacobgkau  would you prefer that I fix that in this PR and make everyone re-approve this, or follow this up with another PR?

**Comment by jacobgkau** [comment]
> would you prefer that I fix that in this PR and make everyone re-approve this, or follow this up with another PR?

Up to you, we're still waiting on engineering approval for https://github.com/pop-os/gnome-shell/pull/82 anyway. If it's something small, it should be pretty quick to re-approve as long as it's not a force push.

**Review by jackpot51 — APPROVED** [review]

**Comment by 13r0ck** [comment]
@jacobgkau f657def should fix that overlap too

**Comment by jacobgkau** [comment]
I saw an issue where if extend the dock to the edges of the screen, set it to always hide, and place it on the bottom, the workspace picker doesn't show up on the right side:

Screenshot from 2022-01-28 15-07-54

I recreated this on two machines (one Intel, one NVIDIA), but after restarting GNOME Shell on both, I'm having trouble getting it to happen again. Trying to confirm if it's a regression or not.

**Review by jacobgkau — APPROVED** [review]
Ok, found the issue & confirmed it's not a regression. Steps to recreate:
1. Set the settings to:
    - Dock always hide
    - Dock extends to edges
    - Dock on the right
    - Workspace picker on the right
2. Restart GNOME Shell
3. Move dock to bottom

Approving this since it's not a regression.

**Pull request merged**
